### ISSUE TITLE
Deleting an LNBank without first deleting the associated bolt cards and withdraw links will result in 500 server errors

### ISSUE THREAD

**Issue by Amperstrand** [dennisreimann/btcpayserver-plugin-lnbank#45]
If you create an LNBank, issue a bolt card and then delete the LNBank, I think you still have the bolt card in the database.

This means that when you visit https://btcpayser/plugins/lnbank/admin/boltcards, you will get a 500 server error because it is trying to list information about bolt cards that are not associated with an LNBank because the LNBank is deleted.

When deleting an LNBank, you should first enforce that all associated Bolt cards are deleted. And also, before deleting a bolt card, you probably want to archive or backup the wipe keys so that it is possible to wipe any cards that have been issued. LNBits does this by forcing you to download a JSON file with the wipe keys when deleting a card. An other option would be to set the status of the card to disabled in the database, but keep the keys. Or maybe deterministic keys.

This issue might not be specific to bolt cards but may also apply to any other LNURLw link attached to an LNBank. Ift he LNBank is deleted, the LNURLw is orphaned, and this may cause 500 server errors. So before deleting LNBanks, enforce that there are no child bolt cards or LNURLw links.

**Comment by dennisreimann** [comment]
Thanks for bringing this up, I'll fix it!

**Comment by dennisreimann** [comment]
Can you check if LNbank v1.8.1 works for you?

**Comment by Amperstrand** [comment]
> Can you check if LNbank v1.8.1 works for you?

At https://btcpayserver/plugins/lnbank/admin, I see previous LNBank wallets that are now deleted. This is good.

At https://btcpayserver/plugins/lnbank/admin/boltcards I get a 500 error. Docker logs show:

```
Microsoft.AspNetCore.Server.Kestrel: Connection id "0HMURP4VV4GO9", Request id "0HMURP4VV4GO9:00000002": An unhandled exception was thrown by the application.
System.NullReferenceException: Object reference not set to an instance of an object.
   at BTCPayServer.Plugins.LNbank.Pages.Admin.Pages_Admin_BoltCards.ExecuteAsync() in /build/BTCPayServer.Plugins.LNbank/Pages/Admin/BoltCards.cshtml:line 26
   at Microsoft.AspNetCore.Mvc.Razor.RazorView.RenderPageCoreAsync(IRazorPage page, ViewContext context)
   at Microsoft.AspNetCore.Mvc.Razor.RazorView.RenderPageAsync(IRazorPage page, ViewContext context, Boolean invokeViewStarts)
   at Microsoft.AspNetCore.Mvc.Razor.RazorView.RenderAsync(ViewContext context)
   at Microsoft.AspNetCore.Mvc.ViewFeatures.ViewExecutor.ExecuteAsync(ViewContext viewContext, String contentType, Nullable`1 statusCode)
   at Microsoft.AspNetCore.Mvc.ViewFeatures.ViewExecutor.ExecuteAsync(ViewContext viewContext, String contentType, Nullable`1 statusCode)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeNextResultFilterAsync>g__Awaited|30_0[TFilter,TFilterAsync](ResourceInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.Rethrow(ResultExecutedContextSealed context)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.ResultNext[TFilter,TFilterAsync](State& next, Scope& scope, Object& state, Boolean& isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.InvokeResultFilters()
--- End of stack trace from previous location ---
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeNextResourceFilter>g__Awaited|25_0(ResourceInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.Rethrow(ResourceExecutedContextSealed context)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.Next(State& next, Scope& scope, Object& state, Boolean& isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeFilterPipelineAsync>g__Awaited|20_0(ResourceInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeAsync>g__Awaited|17_0(ResourceInvoker invoker, Task task, IDisposable scope)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeAsync>g__Awaited|17_0(ResourceInvoker invoker, Task task, IDisposable scope)
   at Microsoft.AspNetCore.Routing.EndpointMiddleware.<Invoke>g__AwaitRequestTask|6_0(Endpoint endpoint, Task requestTask, ILogger logger)
   at Microsoft.AspNetCore.Session.SessionMiddleware.Invoke(HttpContext context)
   at Microsoft.AspNetCore.Session.SessionMiddleware.Invoke(HttpContext context)
   at Microsoft.AspNetCore.Authorization.Policy.AuthorizationMiddlewareResultHandler.HandleAsync(RequestDelegate next, HttpContext context, AuthorizationPolicy policy, PolicyAuthorizationResult authorizeResult)
   at Microsoft.AspNetCore.Authorization.AuthorizationMiddleware.Invoke(HttpContext context)
   at Microsoft.AspNetCore.Authentication.AuthenticationMiddleware.Invoke(HttpContext context)
   at BTCPayServer.Hosting.BTCPayMiddleware.Invoke(HttpContext httpContext) in /source/BTCPayServer/Hosting/BTCpayMiddleware.cs:line 98
   at BTCPayServer.Hosting.GreenfieldMiddleware.Invoke(HttpContext httpContext) in /source/BTCPayServer/Hosting/GreenfieldMiddleware.cs:line 50
   at Microsoft.AspNetCore.Diagnostics.StatusCodePagesMiddleware.Invoke(HttpContext context)
   at BTCPayServer.Hosting.HeadersOverrideMiddleware.Invoke(HttpContext httpContext) in /source/BTCPayServer/Hosting/HeadersOverrideMiddleware.cs:line 30
   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.HttpProtocol.ProcessRequests[TContext](IHttpApplication1 application)
```

I am about 70% sure that I deleted all bolt cards before deleting the parent LNBank wallet, so this error is somewhat unexpected.

Now I have to either wipe the BTCPayServer database and start over or figure out how to delete whatever entries are causing the 500 error.

I don't mind wiping and starting over, but if you have advice for how to work around it or fix the problem, I will leave it for now and document it here if I figure out how to fix it.

**Comment by Amperstrand** [comment]
This might be the card that is causing the issue:

<img width="899" alt="image" src="https://github.com/dennisreimann/btcpayserver-plugin-lnbank/assets/141745238/0642b6f8-6a39-4d39-a7e8-cee6d2d445bf">

It is deleted, but has status "Needs activation". If I click on it, I get a QR-code to activate it. If I visit the activation URL, I get the following message:

```
Object reference not set to an instance of an object.
```

I really like that deleted LNBank instances are still visible. They are not really deleted, but rather disabled. This is even more important for cards: with a previous version of LNBits, I was able to delete the card and the wipe keys and in effect bricking the card. The best UX would be if when you delete a card, the LNURL turns into some sort of message that tells you: "This card has been disabled, the instructions and keys you need to wipe this card can be retrieved by tapping the card 3 times within 60 seconds and visiting the following URL: X

**Comment by dennisreimann** [comment]
> I don't mind wiping and starting over, but if you have advice for how to work around it or fix the problem, I will leave it for now and document it here if I figure out how to fix it.

I will fix it, you don't have to start over.

Also thanks for the rest of the feedback, I'll  take a closer look tomorrow.

**Comment by dennisreimann** [comment]
@Amperstrand I just released LNbank v1.8.2, which should fix the problem.
